// @fluffy-spoon/csharp-parser/dist/src/StructParser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StructParser = void 0;
var Models_1 = require("./Models");
var ScopeHelper_1 = require("./ScopeHelper");
var RegExHelper_1 = require("./RegExHelper");
var MethodParser_1 = require("./MethodParser");
var PropertyParser_1 = require("./PropertyParser");
var FieldParser_1 = require("./FieldParser");
var AttributeParser_1 = require("./AttributeParser");
var StructParser = /** @class */ (function () {
    function StructParser(typeParser) {
        this.typeParser = typeParser;
        this.scopeHelper = new ScopeHelper_1.ScopeHelper();
        this.regexHelper = new RegExHelper_1.RegExHelper();
        this.propertyParser = new PropertyParser_1.PropertyParser();
        this.attributeParser = new AttributeParser_1.AttributeParser();
        this.methodParser = new MethodParser_1.MethodParser(typeParser);
        this.fieldParser = new FieldParser_1.FieldParser(typeParser);
    }
    StructParser.prototype.parseStructs = function (content) {
        var structs = new Array();
        var scopes = this.scopeHelper.getCurlyScopes(content);
        for (var _i = 0, scopes_1 = scopes; _i < scopes_1.length; _i++) {
            var scope = scopes_1[_i];
            var statements = this.scopeHelper.getStatements(scope.prefix);
            for (var _a = 0, statements_1 = statements; _a < statements_1.length; _a++) {
                var statement = statements_1[_a];
                var matches = this.regexHelper.getMatches(statement, new RegExp("^" + this.regexHelper.getStructRegex() + "$", "g"));
                for (var _b = 0, matches_1 = matches; _b < matches_1.length; _b++) {
                    var match = matches_1[_b];
                    try {
                        var attributes = match[0];
                        var modifiers = match[1] || "";
                        var name = match[2];
                        var genericParameters = match[3];
                        var struct = new Models_1.CSharpStruct(name);
                        struct.attributes = this.attributeParser.parseAttributes(attributes);
                        struct.genericParameters = this.typeParser.parseTypesFromGenericParameters(genericParameters);
                        struct.innerScopeText = scope.content;
                        struct.isPublic = modifiers.indexOf("public") > -1;
                        var fields = this.fieldParser.parseFields(scope.content);
                        for (var _c = 0, fields_1 = fields; _c < fields_1.length; _c++) {
                            var field = fields_1[_c];
                            field.parent = struct;
                            struct.fields.push(field);
                        }
                        var properties = this.propertyParser.parseProperties(scope.content);
                        for (var _d = 0, properties_1 = properties; _d < properties_1.length; _d++) {
                            var property = properties_1[_d];
                            property.parent = struct;
                            struct.properties.push(property);
                        }
                        var methods = this.methodParser.parseMethods(scope.content, struct);
                        for (var _e = 0, methods_1 = methods; _e < methods_1.length; _e++) {
                            var method = methods_1[_e];
                            method.parent = struct;
                            struct.methods.push(method);
                        }
                        structs.push(struct);
                    }
                    catch (ex) {
                        console.error("Skipping struct due to parsing error.", statement, ex);
                        debugger;
                    }
                }
            }
        }
        return structs;
    };
    return StructParser;
}());
exports.StructParser = StructParser;
