// @fluffy-spoon/csharp-parser/dist/src/UsingsParser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UsingsParser = void 0;
var Models_1 = require("./Models");
var ScopeHelper_1 = require("./ScopeHelper");
var RegExHelper_1 = require("./RegExHelper");
var UsingsParser = /** @class */ (function () {
    function UsingsParser() {
        this.scopeHelper = new ScopeHelper_1.ScopeHelper();
        this.regexHelper = new RegExHelper_1.RegExHelper();
    }
    UsingsParser.prototype.parseUsings = function (content) {
        var usings = new Array();
        var scopes = this.scopeHelper.getCurlyScopes(content);
        var scope = scopes[0];
        if (!scope)
            return usings;
        var statements = this.scopeHelper.getStatements(scope.prefix);
        for (var _i = 0, statements_1 = statements; _i < statements_1.length; _i++) {
            var statement = statements_1[_i];
            var matches = this.regexHelper.getMatches(statement, new RegExp("^" + this.regexHelper.getUsingRegex(false, true, true) + "$", "g"));
            for (var _a = 0, matches_1 = matches; _a < matches_1.length; _a++) {
                var match = matches_1[_a];
                try {
                    var using = {
                        alias: match[0] || null,
                        namespace: new Models_1.CSharpNamespace(match[1])
                    };
                    usings.push(using);
                }
                catch (ex) {
                    console.error("Skipping using due to parsing error.", statement, ex);
                    debugger;
                }
            }
        }
        return usings;
    };
    return UsingsParser;
}());
exports.UsingsParser = UsingsParser;
